primitives/treasury-funding/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+    type ProvidedUnit: NumericalParam + Into<u64>;
+    /// The ideal system utilization rate. It's a PerThing since it should be a percentage.
+    type IdealUtilizationRate: Get<P>;
+    /// The falloff or decay rate. It's a PerThing since it should be a percentage.
+    type DecayRate: Get<P>;
+    /// The minimum cut for the treasury. It's a PerThing since it should be a percentage.
+    type MinimumCut: Get<P>;
+    /// The maximum cut for the treasury. It's a PerThing since it should be a percentage.
+    type MaximumCut: Get<P>;
+}
+
+impl<T, P> TreasuryCutCalculator for LinearThenPowerOfTwoTreasuryCutCalculator<T, P>
+where
+    T: LinearThenPowerOfTwoTreasuryCutCalculatorConfig<P>,
+    P: PerThing,
+{
+    type Balance = T::Balance;
+    type ProvidedUnit = T::ProvidedUnit;
+    fn calculate_treasury_cut(
+        provided_amount: Self::ProvidedUnit,
+        used_amount: Self::ProvidedUnit,
+        amount_to_charge: Self::Balance,
+    ) -> Self::Balance {
+        // Get the system utilization rate, ideal system utilization rate and falloff from the configuration.
+        let system_utilization =
+            P::from_rational(used_amount.into().into(), provided_amount.into().into());
+        let ideal_system_utilization = T::IdealUtilizationRate::get();
+        let falloff = T::DecayRate::get();
+
+        // Calculate the adjustment to be used to calculate the treasury cut.
+        let adjustment =
+            compute_percentage_to_treasury(system_utilization, ideal_system_utilization, falloff);
+
+        // Get the minimum and maximum cut from the configuration and calculate the difference between them.
+        let minimum_cut = T::MinimumCut::get();
+        let maximum_cut = T::MaximumCut::get();
+        let delta_cut = maximum_cut.saturating_sub(minimum_cut);
+
+        // Calculate the final treasury cut percentage by adjusting the minimum cut with the adjustment times the delta cut.
+        let treasury_cut: FixedU128 = minimum_cut.saturating_add(delta_cut * adjustment).into();
+
+        // Calculate the amount to transfer to the treasury by using the equation `amount_to_charge * treasury_cut_percentage`,
+        // where `treasury_cut_percentage` is obtained by getting the inner value of `treasury_cut` and dividing it by the FixedU128 type
+        // precision. We first multiply the inner value to not end up saturating to 0 when dividing.
+        // Example: if treasury cut ends up being 5%, the inner value of the FixedU128 will be 50_000_000_000_000_000. If the
+        // amount to charge was 10_000_000_000 (one DOT), then the result of this will correctly be 500_000_000 (0.05 DOT).
+        amount_to_charge * treasury_cut.into_inner().into() / FixedU128::DIV.into()
+    }
+}
+
+/// Compute the fraction of charged tokens that go to the treasury using function
+///
+/// ```ignore
+/// I(x) = for x between 0 and x_ideal: 1 - x / x_ideal,
+/// for x between x_ideal and 1: 1 - 2^((x_ideal - x) / d)
+/// ```
```

Comment:
Again, this doesn't actually compute the fraction of charged tokens that go to the treasury.

Reply:
I changed the name and docs to be more precise, LMKWYT